proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,152 @@
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a hash of file content and checking it against a set of known hashes.
+These checks require very little effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effecively every
+client gets "innoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malvare exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are sored on file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the website elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+
+``` idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+    [supportsPromises, permissions="downloads.getFileHash"] static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+``` idl
+namespace downloads {
+  enum FILE_HASH_ALGO {
+    "SHA-256"
+  }
+
+  getFileHash: (downloadId: number, algorithm?: downloads.FILE_HASH_ALGO) => Promise<string>;
+}
+```
+
+### New Permissions
+
+|      Permission Added     | Suggested Warning |
+| ------------------------- | ----------------- |
+| `"downloads.getFileHash"` | `"View cryptographic digests of downloads"` |
+
+The new `"downloads.getFileHash"` permission is optional and is honored only if
+extension has `"downloads"` permission as well.
+
+### Manifest File Changes
+
+No thanges besides the new optional permission `"downloads.getFileHash"`.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+Hashes (cryptographic) are considered one-way operations, but they can easily
+be reversed to the original input when the number of possible inputs is
+sufficiently small. So extension may be able to infer the raw contents of the
+file if the possible space of file contents is small.
+A malicious extension could:
+ - Observe hashes of downloaded files and send them to a home server
```

Comment:
```suggestion
 - obtain the hash of the response to any URL.
```

The downloads API can create downloads without user interaction, so this effectively allows any extension to get the hash for any URL.

Reply:
I added two points:
> - Initiate a download with an arbitrary URL and obtain a hash of the response
> - Initiate a download for a ranged HTTP request and obtain a hash of the
   corresponding portion of the response